includes/class-wpcom-legacy-redirector.php

```diff
@@ -3,513 +3,570 @@
+        $param_values           = array();
+
+        // Parse URL to get Query Params.
+        $query_params = self::mb_parse_url($url, PHP_URL_QUERY);
+        if (! empty($query_params) ) { // Verify Query Params exist.
+
+            // Parse Query String to Associated Array.
+            parse_str($query_params, $param_values);
+            // For every whitelisted param save value and strip from URL.
+            foreach ( $protected_params as $protected_param ) {
+                if (! empty($param_values[ $protected_param ]) ) {
+                    $protected_param_values[ $protected_param ] = $param_values[ $protected_param ];
+                    $url                                        = remove_query_arg($protected_param, $url);
+                }
+            }
+        }
+
+        $url_hash = self::get_url_hash($url);
+
+        $redirect_post_id = wp_cache_get($url_hash, self::CACHE_GROUP);
+
+        if (false === $redirect_post_id ) {
+            $redirect_post_id = self::get_redirect_post_id($url);
+            wp_cache_add($url_hash, $redirect_post_id, self::CACHE_GROUP);
+        }
+
+        if ($redirect_post_id ) {
+            $redirect_post = get_post($redirect_post_id);
+            if (! $redirect_post instanceof WP_Post ) {
+                // If redirect post object doesn't exist, reset cache.
+                wp_cache_set($url_hash, 0, self::CACHE_GROUP);
+
+                return false;
+            } elseif (0 !== $redirect_post->post_parent ) {
+                return add_query_arg($protected_param_values, get_permalink($redirect_post->post_parent)); // Add Whitelisted Params to the Redirect URL.
+            } elseif (! empty($redirect_post->post_excerpt) ) {
+                return add_query_arg($protected_param_values, esc_url_raw($redirect_post->post_excerpt)); // Add Whitelisted Params to the Redirect URL.
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Get Redirect Post ID.
+     *
+     * @param  string $url URL to redirect (source).
+     * @return string|int Redirect post ID (as string) if one was found; otherwise 0.
+     */
+    static function get_redirect_post_id( $url )
+    {
+        global $wpdb;
+
+        $url_hash = self::get_url_hash($url);
+
+        $redirect_post_id = $wpdb->get_var($wpdb->prepare("SELECT ID FROM $wpdb->posts WHERE post_type = %s AND post_name = %s LIMIT 1", self::POST_TYPE, $url_hash));
+
+        if (! $redirect_post_id ) {
+            $redirect_post_id = 0;
+        }
+
+        return $redirect_post_id;
+    }
+
+    /**
+     * Utility to get MD5 hash of URL.
+     *
+     * @param  string $url URL to hash.
+     * @return string Hash representation of string.
+     */
+    private static function get_url_hash( $url )
+    {
+        return md5($url);
+    }
+
+    /**
+     * UTF-8 aware parse_url() replacement.
+     * 
+     * @return array
+     */
+    public static function mb_parse_url($url, $component = -1)
```

Comment:
This whole method should be written using the WP coding style - lots of whitespace, particularly inside `( )`.